### PR TITLE
DOC: interpolate: Extrapolation tips and tricks

### LINKED ISSUE #14386: Let CloughTocher2DInterpolator fit "nearest" for points outside the convex hull (instead of nans)
Currently, [CloughTocher2DInterpolator](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.CloughTocher2DInterpolator.html#scipy.interpolate.CloughTocher2DInterpolator) can only fill in a constant value (nan by default) when queried outside the convex hull of the input points. But this absolutely isn't ideal.

What would be much better is to use the value of the nearest neighbor, which [NearestNDInterpolator](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.NearestNDInterpolator.html#scipy.interpolate.NearestNDInterpolator) does by default. For CloughTocher2DInterpolator, which already constructs the Delaunay Triangulation of the input points, finding the nearest point on the convex hull should be efficient.

Here's just my naive implementation for filling in the nearest neighbor value outside the convex hull of the inputs.

```python3
import matplotlib.pylab as P
from scipy.interpolate import CloughTocher2DInterpolator as CT
def my_CT(xy, z):
  x = xy[:, 0]
  y = xy[:, 1]
  f = CT(xy, z)
  def new_f(xx, yy):
    xx = P.array(xx) + [0.0]
    yy = P.array(yy) + [0.0]
    zz = f(xx, yy)
    nan = P.isnan(zz)
    if nan.any():
      inds = P.argmin(
          (x.reshape(-1, 1) - xx[nan])**2 + 
          (y.reshape(-1, 1) - yy[nan])**2
          , 0)
      zz[nan] = z[inds]
    return zz
  return new_f
```

### LINKED ISSUE #14472: scipy.interpolate.CubicSpline boundary conditions appear to be incorrect
scipy.interpolate.CubicSpline allows for a series of boundary conditions in bc_type. These do not appear to be respected.

For example, 'natural' should result in second derivative being zero outside the x-values of the data, whilst 'clamped' should result in the first derivative being zero outside this range.

There is some difference between the splines generated with each setting of boundary condition, but the derivatives do not match the description.

#### Reproducing code example:
```
import matplotlib.pyplot as plt

import numpy as np

from scipy.interpolate import CubicSpline

xs = [1,2,3,4,5,6,7,8]
ys = [4.5, 3.6, 1.6, 0.0, -3.3, -3.1, -1.8, -1.7]

notaknot_smoother = CubicSpline(xs,ys, bc_type='not-a-knot')

natural_smoother = CubicSpline(xs,ys, bc_type='natural')

clamped_smoother = CubicSpline(xs,ys, bc_type='clamped')

extrapolated = np.linspace(min(xs)-10,max(xs)+10,int(21+(max(xs)-min(xs))))

fig, axs = plt.subplots(3, 3,figsize=(12,12))
axs[0, 0].plot(xs, ys,'.')
axs[0, 0].plot(extrapolated,notaknot_smoother(extrapolated),'-')
axs[0, 0].set_title('not-a-knot, deriv=0')
axs[0, 1].plot(xs, ys,'.')
axs[0, 1].plot(extrapolated,natural_smoother(extrapolated),'-')
axs[0, 1].set_title('natural, deriv=0')
axs[0, 2].plot(xs, ys,'.')
axs[0, 2].plot(extrapolated,clamped_smoother(extrapolated),'-')
axs[0, 2].set_title('clamped, deriv=0')
axs[1, 0].plot(xs, notaknot_smoother(xs,nu=1),'.')
axs[1, 0].plot(extrapolated,notaknot_smoother(extrapolated,nu=1),'-')
axs[1, 0].set_title('not-a-knot, deriv=1')
axs[1, 1].plot(xs, natural_smoother(xs,nu=1),'.')
axs[1, 1].plot(extrapolated,natural_smoother(extrapolated,nu=1),'-')
axs[1, 1].set_title('natural, deriv=1')
axs[1, 2].plot(xs, clamped_smoother(xs,nu=1),'.')
axs[1, 2].plot(extrapolated,clamped_smoother(extrapolated,nu=1),'-')
axs[1, 2].set_title('clamped, deriv=1')
axs[2, 0].plot(xs, notaknot_smoother(xs,nu=2),'.')
axs[2, 0].plot(extrapolated,notaknot_smoother(extrapolated,nu=2),'-')
axs[2, 0].set_title('not-a-knot, deriv=2')
axs[2, 1].plot(xs, natural_smoother(xs,nu=2),'.')
axs[2, 1].plot(extrapolated,natural_smoother(extrapolated,nu=2),'-')
axs[2, 1].set_title('natural, deriv=2')
axs[2, 2].plot(xs, clamped_smoother(xs,nu=2),'.')
axs[2, 2].plot(extrapolated,clamped_smoother(extrapolated,nu=2),'-')
axs[2, 2].set_title('clamped, deriv=2')
plt.show()
```

#### Scipy/Numpy/Python version information:

Confirmed in both these versions:

1.5.4 1.19.5 sys.version_info(major=3, minor=6, micro=7, releaselevel='final', serial=0)

1.7.0 1.21.1 sys.version_info(major=3, minor=8, micro=0, releaselevel='final', serial=0)

### LINKED ISSUE #16338: ENH: Add option to clip out of bounds input values to minimum and maximum of search array in scipy.interpolate.interp1d()
### Is your feature request related to a problem? Please describe.

[scipy.interpolate.interp1d](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp1d.html) currently only allows `“extrapolate”` as method to fill values. If you want to define a tuple with the search bounds, you have to calculate those yourself.

### Describe the solution you'd like.

I would like to propose a new built-in method to fill values, which clips the return values to the minimum and maximum of your search array. This way it can only return values at the bounds of your search array, this would be equivalent to `fill_value=(min(search_array), max(search_array)), bounds_error=False`.

It could be called `"clip"` and used as `fill_value="clip"`.

### Describe alternatives you've considered.

Of course it's not that cumbersome to implement it yourself, but I thought it would be a nice quality-of-life addition. It can also also potentially reach higher performance with a built-in implementation.

### Additional context (e.g. screenshots, GIFs)

_No response_

### PR DESCRIPTION
**Pull request by ev-br**
#### Reference issue

closes gh-16338, closes gh-14386, closes gh-14472

#### What does this implement/fix?

Add a tutorial page with several worked examples of extrapolation--- handling out-of-bounds data --- in several interpolators. In general, extrapolation is not entirely consistent in scpy.interpolate, default modes are not always useful, and in some cases manual work on the user side is needed. There are four examples, of both using the existing keyword switches etc, and DIY implementations. 

#### Additional information

Two out of four examples are provided by users, in gh-14386 and gh-14472. Am crediting the users in the commit messages, will comment on the issues to see if they want to be listed as commit co-authors.

### CODE REVIEW

**Comment by ev-br** [comment]
CI failures are unrelated, as it is a pure DOC change.

File: doc/source/tutorial/interpolate/extrapolation_examples.rst [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+            axs[i, j].plot(xs, spline(xs, nu=i),'o')
+            axs[i, j].plot(xnew, spline(xnew, nu=i),'-')
+            axs[i, j].set_title(f'{title}, deriv={i}')
+            
+    plt.tight_layout()
+    plt.show()
+
+It is clearly seen that e.g. natural spline does have the zero second derivative
+at the boundaries, but extrapolation is non-linear. ``bc_type="clamped"``
+shows a similar behavior: first derivatives are only equal zero exactly at the
```
Comment (mdhaber) [inline doc/source/tutorial/interpolate/extrapolation_examples.rst:143]:
```suggestion
shows a similar behavior: first derivatives are only equal to zero exactly at the
```

File: doc/source/tutorial/interpolate/extrapolation_examples.rst [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+boundary. In all cases, extrapolation is done by extending the first and last
+polynomial pieces of the spline, whatever they happen to be.
+
+One possible way to force the extrapolation is to extend the interpolation domain
+to add a first and last polynomial pieces which have desired properties.
```
Comment (mdhaber) [inline doc/source/tutorial/interpolate/extrapolation_examples.rst:148]:
```suggestion
to add first and last polynomial pieces which have desired properties.
```
And I'd prefer "polynomial pieces with the desired properties."

**Review by mdhaber — COMMENTED** [review]
Since this is all new and the information is correct/useful, I think we can merge this pretty quickly. 

I agree that this closes the issues if you think it would make the API too messy to include options for these things. 

I've only made it halfway through. Next pass I'll start from "Manually implement the asymptotic".

File: doc/source/tutorial/interpolate.rst [hunk 3]
```diff
@@ -30,6 +30,11 @@ for data in 1, 2, and higher dimensions:
 
 - Interpolation using radial basis functions.
 
+.. toctree::
+   :maxdepth: 3
+
+   interpolate/extrapolation_examples
+
```
[the remark this answers is not in the document]
Comment (ev-br) [reply doc/source/tutorial/interpolate.rst]: Uhm. I don't believe these should be two equally sized top-level headings. In the last commit this page is listed along the rest of interpolate tutorial pages. Which is a tad strange indeed. 

Originally, this page was intended to be a collection of assorted recipes, really. Not a full-fledged discussion of how different interpolators do it (`fill_value="extrapolate, bounds_error=False"` or `extrapolate=True/False/None` etc). The idea was that these details are in the API docs and this page is something to consult when a user wants something unusual. I can expand this page with a summary (a table? a listing?) of what various objects do if you think duplicating it from the API docs is useful?

File: doc/source/tutorial/interpolate/extrapolation_examples.rst [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+``xnew < x[0]`` and ``y[-1]`` for ``xnew > x[-1]``.
+
+By default, `interp1d` refuses to extrapolate, and raises a
+``ValueError`` when evaluated on a data point outside of the
+interpolation range. This can be switched off by the
+``bounds_error=False`` argument: then ``interp1d`` sets the out-of-range
+values with the ``fill_value``, which is ``nan`` by default.
+
+To mimic the behavior of `numpy.interp` with `interp1d`, you can use
+the fact that it supports a 2-tuple as the ``fill_value``. The tuple
```
[the remark this answers is not in the document]
Comment (ev-br) [reply doc/source/tutorial/interpolate/extrapolation_examples.rst:52]: Unless this page is expanded to cover the options of various interpolators, not really, no :-).

**Comment by ev-br** [comment]
I think the last commit addresses all review suggestions.

File: doc/source/tutorial/interpolate/extrapolation_examples.rst [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+
+However, if we need to solve it multiple times (e.g. to find *a series*
+of roots due to periodicity of the ``tan`` function), repeated calls to
+`scipy.optimize.brentq` become prohibitively expensive.
+
+To circumvent this difficulty, we tabulate :math:`y = ax - 1/\tan{x}`
+and interpolate it on the tabulated grid. In fact, we will use the *inverse*
+interpolation: we interpolate the values of :math:`x` versus :math:`у`.
+This way, solving the original equation becomes simply an evaluation of
+the interpolated function at zero argument.
```
Comment (j-bowhay) [inline doc/source/tutorial/interpolate/extrapolation_examples.rst:282]: To me here it is not clear what argument is referring to

File: doc/source/tutorial/interpolate/extrapolation_examples.rst [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+    for i in [0, 1, 2]:
+        for j, spline, title in zip(range(3), splines, titles):
+            axs[i, j].plot(xs, spline(xs, nu=i),'o')
+            axs[i, j].plot(xnew, spline(xnew, nu=i),'-')
+            axs[i, j].set_title(f'{title}, deriv={i}')
+            
+    plt.tight_layout()
+    plt.show()
+
+It is clearly seen that e.g. natural spline does have the zero second
```
Comment (j-bowhay) [inline doc/source/tutorial/interpolate/extrapolation_examples.rst:146]: This reads like there is something missing from before the eg?

**Review by j-bowhay — COMMENTED** [review]
A few minor comments

**Comment by ev-br** [comment]
Thanks Jake for the review, I tried to address your comments in the last commit in this branch.

**Comment by mdhaber** [comment]
I wanted to finish my review, the rendered artifacts are not available right now for some reason (looks like a CircleCI issue. Since I'm not very familiar with these functions, the rendered documentation is important, so I'll have to come back. I'll try to look at the rest (starting at "Manually Implement the Asymptotics") on Thursday.

**Review by j-bowhay — APPROVED** [review]
Thanks @ev-br that is my comments addressed

File: doc/source/tutorial/interpolate.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdhaber) [reply doc/source/tutorial/interpolate.rst]: This looks better now!

**Comment by mdhaber** [comment]
My original comments have been resolved, I don't have any new ones, and this is very useful. Thanks @ev-br @j-bowhay!

**Comment by mdhaber** [comment]
Oh, the only question is what we should do about co-authorship of @chausies. I see you asked whether they'd like to be credited, but there is no response. What would you like to do? Besides https://github.com/scipy/scipy/issues/14472#issuecomment-1294665827, you can usually find email addresses by adding `.patch` to the URL of any commit. e.g., https://github.com/chausies/myconfigs/commit/b33ded48276f9bcdfcee667c4b7c9dad783cc9e0.patch .

**Comment by ev-br** [comment]
I think the lack of response on GH for a reasonable time can be interpreted as a lazy agreement to the attribution in the commit message of https://github.com/scipy/scipy/pull/17299/commits/4237531b7d745368aaf6b761a4359c80ed6762b8 --- this used the information which is already made public by @chausies. 

The comment on gh-14386 has two weeks, we can probably wait some more.

**Comment by ev-br** [comment]
Thanks Matt, Jake for the reviews!

**Pull request merged**